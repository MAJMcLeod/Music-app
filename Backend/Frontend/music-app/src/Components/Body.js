import './Style/Body.css'
import { useState, useEffect } from 'react'
import ListContainer from './ListContainer'
import SearchContainer from './SearchContainer'
import useAuth from './UseAuth'
import Nav from './Nav'
import SpotifyWebApi from 'spotify-web-api-node'
//importing appropriate react components and stylesheets

const spotifyApi = new SpotifyWebApi({
    clientId: '1d0da610ffee473cadc359df2f1e06f8'
})
//Initialising a new SpotifyApi component and giving it a unique client id  
// in order to verify identity

let Body = ({ code }) => {
    // use of hooks to pass through the access token recieved by the parent component
    const accessToken = useAuth(code)
    const [search, setSearch] = useState('')
    const [results, setResults] = useState([])
    const [selectedTrack, setSelectedTrack] = useState()
    const [favourites, setFavourites] = useState(JSON.parse(sessionStorage.getItem('favItems')))
    // use of useState to store information and update the Body component when information is changed

    function addFavourite(track) {
        window.sessionStorage.setItem("favItems", JSON.stringify([...favourites, track]));
        setFavourites([...favourites, track])
    }
    // function that takes in a track object and saves it to session storage and favourites array along
    // with all previous entries in the previous favourites array

    function removeFavourite(key) {
        window.sessionStorage.setItem("favItems", JSON.stringify(favourites.filter(track => track.uri !== key.uri)));
        setFavourites(favourites.filter(track => track.uri !== key.uri));
    }
    // function that takes in a track(key) and filters the favourites array by comparing each entries uri to the track's uri
    // and removing whichever entry matches the uri. It then updates the array and saves it to session storage and favourites array

    function selectTrack(track) {
        setSelectedTrack(track)
    }
    // function that takes in a track and updates the selectTrack useState

    useEffect(() => {
        // useEffect executes everytime the component updates
        if (!accessToken) return
        // if no access token is present then the code will stop here
        spotifyApi.setAccessToken(accessToken)
        // sets the access token for the spotifyApi so that it can be used
    }, [accessToken])

    useEffect(() => {
        if (!search) return setResults([])
        // if there is nothing in search then the function will stop and make
        // the results array empty
        if (!accessToken) return
        // if no access token is present then the code will stop here

        spotifyApi.searchTracks(search).then(res => {
            // search function of the spotifyApi in which you pass a string and
            // it searches accordingly

            setResults(res.body.tracks.items.map(track => {
                // takes the result from the Api and maps out the items into
                // track objects that are then stored in results for rendering
                return {
                    artist: track.artists[0].name,
                    uri: track.uri,
                    song: track.name,
                    albumCover: track.album.images[0].url
                    // extracting the appropriate data from the results
                    // and storing them 
                }
            }))
        })
    }, [search, accessToken])

    let SearchCallback = (searchInput) => {
        setSearch(searchInput);
    }
    // callback function that takes the search string from the search child component
    // and saves it in the search useState

    return (
        // returning elements to render to the DOM
        <div className="Body">
            <Nav accessToken={accessToken} trackUri={selectedTrack?.uri} />
            {/* passing data via props to child elements for use in functions */}

            <div className="BodyInner">
                <SearchContainer SearchCallback={SearchCallback} removeFavourite={removeFavourite} />
                {/* passing functions via props to child elements for use in callbacks */}
                <ListContainer results={results} selectTrack={selectTrack} addFavourite={addFavourite} />
            </div>

            <div className="Background">
                <div className="Pattern">
                </div>
            </div>
        </div>
    )
}

export default Body;
// exporting the component for use in other react components