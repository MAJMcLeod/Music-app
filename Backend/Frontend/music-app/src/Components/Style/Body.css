.Body {
    display: flex;
    justify-content: space-around;
    width: 100%;
    height: 700px;
}

.BodyInner {
    height: 110%;
    display: flex;
    width: 80%;

}

.Background {
    background-color: #06001a;;
    z-index: -1;
    position: fixed;
    width: 100%;
    height: 100vh;



}

