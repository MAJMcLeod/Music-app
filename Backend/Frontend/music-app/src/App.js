import './App.css';
import Body from './Components/Body';
import Login from './Components/Login'
//importing appropriate react components and stylesheets

const code = new URLSearchParams(window.location.search).get('code');
// variable used to store the code that is sent by the API via the 
// http string

if (!JSON.parse(sessionStorage.getItem('favItems'))) {
  window.sessionStorage.setItem("favItems", JSON.stringify([]));
}
// if there are no items in session storage then it is set to the default
// value of an empty array so that it is never truly empty, stopping many 
// future errors 

function App() {

  return (
    <div className="App">
      {code
        ? <Body code={code} />
        : <Login />}
        {/* conditional render statement */}
    </div>
  );
}

export default App;
