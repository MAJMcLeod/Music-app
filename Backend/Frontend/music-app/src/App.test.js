import { render, screen } from '@testing-library/react';
import App from './App';
import Body from './Components/Body'
import renderer from 'react-test-renderer'

describe("Snapshot Test", () => {
  test("Snapshot for App.js Component", () => {
    const renderedApp= renderer.create(<App/>).toJSON()
    expect(renderedApp).toMatchSnapshot();
  })

  test("Snapshot for Body.js Component", () => {
    const renderedApp= renderer.create(<Body/>).toJSON()
    expect(renderedApp).toMatchSnapshot();
  })
})
// tests that check to see if each respective component matches the 
// saved snapshots
