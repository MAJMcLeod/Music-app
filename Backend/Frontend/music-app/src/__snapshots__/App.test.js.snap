// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test Snapshot for App.js Component 1`] = `
<div
  className="App"
>
  <div
    className="Login"
  >
    <a
      className="LoginLink"
      href="https://accounts.spotify.com/authorize?client_id=1d0da610ffee473cadc359df2f1e06f8&response_type=code&redirect_uri=http://localhost:3000/&scope=streaming%20user-read-email%20user-read-private%20user-library-read%20user-library-modify%20user-read-playback-state%20user-modify-playback-state"
    >
      Login With Spotify
    </a>
  </div>
</div>
`;

exports[`Snapshot Test Snapshot for Body.js Component 1`] = `
<div
  className="Body"
>
  <div
    className="BodyInner"
  >
    <div
      className="SearchHolder"
    >
      <p
        className="Title"
      >
        Music-App
      </p>
      <div
        className="SearchContainer"
      >
        <div
          className="SearchOptions"
        >
          <input
            className="SearchBar"
            onChange={[Function]}
            placeholder="Search"
          />
          <div
            className="SearchIconDiv"
            onClick={[Function]}
          >
            <svg
              className="Search"
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="11"
                cy="11"
                r="8"
              />
              <line
                x1="21"
                x2="16.65"
                y1="21"
                y2="16.65"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        className="FavouritesOuter"
      >
        <div
          className="Favourites"
        >
          <div
            className="FavSpacer"
          >
            <p
              className="FavText"
            >
              Favourites
            </p>
          </div>
          <div
            className="FavouritesInner"
          >
            <div
              className="FavContainer"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="ListContainer"
    >
      <div
        className="ListContainerOuter"
      >
        <div
          className="ListContainerList"
        >
          <div
            className="Spacer"
          />
          <div
            className="Spacer2"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="Background"
  >
    <div
      className="Pattern"
    />
  </div>
</div>
`;
